beacon-chain/blockchain/process_block.go

```diff
@@ -126,11 +114,23 @@ func (s *Service) postBlockProcess(ctx context.Context, signed interfaces.ReadOn
 	})
 
 	defer reportAttestationInclusion(b)
-	//only handle epoch boundary if the incoming block is canonical,
-	//otherwise this will be handled by lateBlockTasks.
 	if headRoot == blockRoot {
-		if err := s.handleEpochBoundary(ctx, postState.Slot(), postState, blockRoot[:]); err != nil {
-			return errors.Wrap(err, "could not handle epoch boundary")
+		// Updating next slot state cache can happen in the background
+		// except in the epoch boundary in which case we lock to handle
+		// the shuffling and proposer caches updates.
+		// We handle these caches only on canonical
+		// blocks, otherwise this will be handled by lateBlockTasks
+		slot := postState.Slot()
+		if (slot+1)%params.BeaconConfig().SlotsPerEpoch != 0 {
+			go updateNextSlotCacheOnBlock(blockRoot, postState)
+		} else {
+			if err := transition.UpdateNextSlotCache(ctx, blockRoot[:], postState); err != nil {
+				return errors.Wrap(err, "could not update next slot state cache")
+			}
+			if err := s.handleEpochBoundary(ctx, slot, postState, blockRoot[:]); err != nil {
+				return errors.Wrap(err, "could not handle epoch boundary")
```

Comment:
Do we really need to return errors here? I know this is an old behavior

Reply:
I don't know this one actually.